### PR TITLE
Remember important last used reach values

### PR DESCRIPTION
**Pull request by m-kuhn**
The add reach wizard remembers the last used values, as specified in https://github.com/QGEP/QGEP/issues/163

@urskaufmann as discussed

### CODE REVIEW

**Comment by urskaufmann** [comment]
Nearly perfect!  Only the ch_function_hydraulic - value is not reused.

**Comment by urskaufmann** [comment]
Bug with reachpoint_level: if there is no snapping (not to node and not to other reach), I get a level of 4000 m.a.s.l.

File: src/tools/qgepmaptooladdfeature.py [hunk 1]
```diff
@@ -303,13 +304,18 @@ def right_clicked(self, _):
...
             f = QgsFeature(fields)
-            for idx in range(len(fields)):
-                # try client side default value first
-                v = self.layer.defaultValue(idx, f)
-                if v != NULL:
-                    f.setAttribute(idx, v)
+            if not self.last_feature_attributes:
+                self.last_feature_attributes = [None] * fields.count()
+            for idx, field in enumerate(fields):
+                if field.name() in ['clear_height', 'material', 'ch_usage_current', 'ch_function_hierarchic', 'horizontal_positioning', 'ws_status', 'ws_year_of_construction', 'ws_fk_owner', 'ws_fk_operator', 'inside_coating', 'fk_pipe_profile', 'remark']:
```
Comment (m-kuhn) [inline src/tools/qgepmaptooladdfeature.py]:
```suggestion
                if field.name() in ['clear_height', 'material', 'ch_usage_current', 'ch_function_hierarchic', 'ch_function_hydraulic','horizontal_positioning', 'ws_status', 'ws_year_of_construction', 'ws_fk_owner', 'ws_fk_operator', 'inside_coating', 'fk_pipe_profile', 'remark']:
```

**Comment by m-kuhn** [comment]
> Nearly perfect! Only the ch_function_hydraulic - value is not reused.
Done

> Bug with reachpoint_level: if there is no snapping (not to node and not to other reach), I get a level of 4000 m.a.s.l.

Is this new or already before?

**Comment by urskaufmann** [comment]
I think it was before. I think it is since the z-geometry synchronisation.  If the reachpoint snaps, I get the level of the snapped object or NULL. If there is no snap, I get 4000.

**Comment by m-kuhn** [comment]
@signedav any idea why that could be?

**Comment by signedav** [comment]
I don't really know. The expression for the default value is this:

**Comment by signedav** [comment]
> I receive NULL if no snapping happens.
No. I receive 0. This should not happen neither I guess...

**Comment by m-kuhn** [comment]
Thanks @signedav 

@urskaufmann did you change the default z value in options/digitizing

**Comment by urskaufmann** [comment]
I did not change. Its the demoproject, that has this expression:
vw_qgep_reach

**Comment by urskaufmann** [comment]
And no changes in Options/Digitizing:
optiondigitizing

**Comment by m-kuhn** [comment]
> I think it was before. I think it is since the z-geometry synchronisation. If the reachpoint snaps, I get the level of the snapped object or NULL. If there is no snap, I get 4000.

Looks like an unrelated issue in this case, maybe even local because

> I cannot reproduce it at the moment. I receive NULL if no snapping happens.

... so let's merge this pull request and rollout the fix and in parallel open a new issue. It would be good to tell us in the new issue if it also happens on a clean new QGIS profile and if others are experiencing the same.

**Pull request merged**